### ISSUE TITLE
index field: "extension"

### ISSUE THREAD

**Issue by ccev** [UIcons/UIcons#11]
### Importance: essential
- Type: `xxx` (defined per-folder)

This is the file ending of all icons in that folder. So we could not only support PNGs but also GIFs, WEBPs or anything else.

EDIT: I was originally suggesting `file_ending` but changed it to `extension` as suggested by versx. 

```json
{
    "pokemon": {
        "extension": "gif"
    },
    "pokestop": {
        "extension": "png"
    }
}
```

**Comment by nileplumb** [comment]
Would this be to have all types within the same directory eg Pokémon webp and Pokémon PNG or would that still be best in it's own parent project directory

**Comment by ccev** [comment]
Nor sure what you mean. This would allow you to have GIF Pokemon and PNG Pokestops. But not a GIF Bulbasaur and a PNG Charmander

**Comment by versx** [comment]
I suggest either `ext` or `extension`
```
{
    "pokemon": {
        "ext": "gif"
    },
    "pokestop": {
        "extension": "png"
    }
}
```

**Comment by ccev** [comment]
`extension` would definitely be the better name here 👍

**Comment by nileplumb** [comment]
> Nor sure what you mean. This would allow you to have GIF Pokemon and PNG Pokestops. But not a GIF Bulbasaur and a PNG Charmander

I quite often get DM'd to make webp versions of my repo's for telegram, so didn't know if this would allow for both in the same parent folder or if better off as I currently have eg home,home_outlined,home_webp

**Comment by ccev** [comment]
>  or if better off as I currently have eg home,home_outlined,home_webp

That's how you'd have to do it

**Comment by versx** [comment]
> > or if better off as I currently have eg home,home_outlined,home_webp
> 
> That's how you'd have to do it

🤔 Wouldn't this PR allow `png` and `webp` files in the same `/pokemon` folder (as well as `jpg`, `gif`, etc)? Unless I'm misunderstanding something?

**Comment by ccev** [comment]
Nope, since the extension is defined within the `pokemon` key of the index json. So it'd only allow exactly one extension per icontype. anything else would be too complicated imo.

Making a separate webp repo would be the way to go here.

I was thinking of a possible field to allow defining a custom fallback repo. Decided to scrap that idea. Though it'd definitely be useful in this situation

**Comment by jepke** [comment]
If every extension will get its own repo what's the advantage of listing filenames in the json without extensions? Makes the json bigger by a couple bytes and requires more lines of code to use it.

**Comment by ccev** [comment]
I'm not sure if you understand my point. Defining the extension for each name in the `icons` list would make the json bigger.

Having the extension defined individually also makes it easier and more flexible to code.

Maybe I'm not understanding you correctly. But with the way UIcons work, there's no way to support a full png and webp pokemon iconset in one repo. 

I also don't think it's a good idea to allow different file types per icon type as it would mess with tools.

**Comment by versx** [comment]
> Nope, since the extension is defined within the `pokemon` key of the index json. So it'd only allow exactly one extension per icontype. anything else would be too complicated imo.
> 
> Making a separate webp repo would be the way to go here.
> 
> I was thinking of a possible field to allow defining a custom fallback repo. Decided to scrap that idea. Though it'd definitely be useful in this situation

I meant within the same repository, yes only one would be selected of course based on whatever is set for extension.

**Comment by jepke** [comment]
> I'm not sure if you understand my point. Defining the extension for each name in the `icons` list would make the json bigger.
> 
> Having the extension defined individually also makes it easier and more flexible to code.
> 
> Maybe I'm not understanding you correctly. But with the way UIcons work, there's no way to support a full png and webp pokemon iconset in one repo.
> 
> I also don't think it's a good idea to allow different file types per icon type as it would mess with tools.

How will you handle if png and webp are I one repo. But icon 3_e_f1234_c12 is only available for png. This way you need 2 separate index.json. Might as well just say 1 supported file extension per repo. 

Yes this can be solved in many ways but just keep it simple 1 repo 1 single format so png and webp both get there own repo.

**Comment by ccev** [comment]
What you mentioned is the way to do it imo. And it's what my suggestion would support.

However, I'd like to support repos that e.g. provide GIF pokemon icons but PNG stop icons. Since GIF stop  icons don't really make sense, while GIF mons do make sense

**Comment by versx** [comment]
> Nope, since the extension is defined within the `pokemon` key of the index json. So it'd only allow exactly one extension per icontype. anything else would be too complicated imo.
Then there's no reason to specify the extension outside of the actual list of available icons, period.

> Making a separate webp repo would be the way to go here.
> 
> I was thinking of a possible field to allow defining a custom fallback repo. Decided to scrap that idea. Though it'd definitely be useful in this situation

Useless

**Comment by versx** [comment]
> > Nope, since the extension is defined within the `pokemon` key of the index json. So it'd only allow exactly one extension per icontype. anything else would be too complicated imo.  

Then there's no reason to specify the extension outside of the actual list of available icons, period.

> > Making a separate webp repo would be the way to go here.
> > I was thinking of a possible field to allow defining a custom fallback repo. Decided to scrap that idea. Though it'd definitely be useful in this situation  

Useless
